client/consensus/src/lib.rs

```diff
@@ -245,7 +245,147 @@ where
                 None
             }
         } else {
+            log::debug!("Invalid nimbus id: {:?}", type_public_pair);
+
             None
         }
     })
 }
+
+/// Grab the first eligible nimbus key from the keystore
+/// If multiple keys are eligible this function still only returns one
+/// and makes no guarantees which one as that depends on the keystore's iterator behavior.
+/// This is the standard way of determining which key to author with.
+/// It also returns its ParaId assignment
+pub async fn first_eligible_key_solochain<B: BlockT, C, P>(
+    client: &C,
+    parent_hash: &H256,
+    keystore: KeystorePtr,
+) -> Option<(AuthorityId<P>, ParaId)>
+where
+    C: RelayChainInterface + ?Sized,
+    P: Pair + Send + Sync,
+    P::Public: AppPublic + Hash + Member + Encode + Decode,
+    P::Signature: TryFrom<Vec<u8>> + Hash + Member + Encode + Decode,
+    AuthorityId<P>: From<<NimbusPair as sp_application_crypto::Pair>::Public>,
+{
+    // Get all the available keys
+    let available_keys = Keystore::keys(&*keystore, NIMBUS_KEY_ID).ok()?;
+
+    // Print a more helpful message than "not eligible" when there are no keys at all.
+    if available_keys.is_empty() {
+        log::warn!(
+            target: LOG_TARGET,
+            "🔏 No Nimbus keys available. We will not be able to author."
+        );
+        return None;
+    }
+
+    // Iterate keys until we find an eligible one, or run out of candidates.
+    for type_public_pair in available_keys {
+        if let Ok(nimbus_id) = NimbusId::from_slice(&type_public_pair) {
+            // If we dont find any parachain that we are assigned to, return none
+            if let Ok(Some(para_id)) =
+                solochain_check_para_id_assignment(client, parent_hash, nimbus_id.clone()).await
+            {
+                log::debug!("Para id found for assignment {:?}", para_id);
+
+                return Some((nimbus_id.into(), para_id));
+            } else {
+                log::debug!("No Para id found for assignment {:?}", nimbus_id);
+            }
+        } else {
+            log::debug!("Invalid nimbus id: {:?}", type_public_pair);
+        }
+    }
+
+    None
+}
+
+/// Grab the first eligible nimbus key from the keystore
+/// If multiple keys are eligible this function still only returns one
+/// and makes no guarantees which one as that depends on the keystore's iterator behavior.
+/// This is the standard way of determining which key to author with.
+/// It also returns its ParaId assignment
+pub async fn first_eligible_key_next_session_solochain<B: BlockT, C, P>(
+    client: &C,
+    parent_hash: &H256,
+    keystore: KeystorePtr,
+) -> Option<(AuthorityId<P>, ParaId)>
+where
+    C: RelayChainInterface + ?Sized,
```

Comment:
The only thing that changes here with respect to the previous function is the bounds on C?

Reply:
The bounds on C and this one is async while the old one is not. So we can't use `find_map` here. If we wanted to unify them we would have to make them both async, I can try.

Reply:
that would be great

Reply:
I can remove this function if we add some methods to the OrchestratorChainInterface: https://github.com/moondance-labs/dancekit/pull/33

Reply:
*fixed*